encoding/thrift/thriftrw-plugin-yarpc/golden_test.go

```diff
@@ -0,0 +1,177 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main_test
+
+import (
+	"crypto/sha1"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// This implements a test that verifies that the code in testdata/ is up to
+// date.
+
+const _testPackage = "go.uber.org/yarpc/encoding/thrift/thriftrw-plugin-yarpc/testdata"
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+
+	// We put the current version of the plugin on the path first.
+	outputDir, err := ioutil.TempDir("", "current-thriftrw-plugin-yarpc")
+	if err != nil {
+		log.Fatalf("failed to create temporary directory: %v", err)
+	}
+	defer os.RemoveAll(outputDir)
+
+	path := os.Getenv("PATH")
+	if err := os.Setenv("PATH", fmt.Sprintf("%v:%v", outputDir, path)); err != nil {
+		log.Fatalf("failed to add %q to PATH: %v", outputDir, err)
+	}
+
+	cmd := exec.Command(
+		"go", "build", "-o", filepath.Join(outputDir, "thriftrw-plugin-yarpc"), ".")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		log.Fatalf("failed to build plugin: %v", err)
+	}
+
+	os.Exit(m.Run())
+}
+
+func TestCodeIsUpToDate(t *testing.T) {
+	thriftRoot, err := filepath.Abs("testdata")
+	require.NoError(t, err, "could not resolve absolute path to testdata")
+
+	thriftFiles, err := filepath.Glob(thriftRoot + "/*.thrift")
+	require.NoError(t, err)
+
+	outputDir, err := ioutil.TempDir("", "golden-test")
+	require.NoError(t, err, "failed to create temporary directory")
+	defer os.RemoveAll(outputDir)
+
+	// TODO: compile plugin and put it on the path
+
+	for _, thriftFile := range thriftFiles {
+		packageName := strings.TrimSuffix(filepath.Base(thriftFile), ".thrift")
+		currentPackageDir := filepath.Join("testdata", packageName)
+		newPackageDir := filepath.Join(outputDir, packageName)
+
+		currentHash, err := dirhash(currentPackageDir)
+		require.NoError(t, err, "could not hash %q", currentPackageDir)
+
+		err = thriftrw(
+			"--no-recurse",
+			"--out", outputDir,
+			"--pkg-prefix", _testPackage,
+			"--thrift-root", thriftRoot,
+			"--plugin", "yarpc",
+			thriftFile,
+		)
+		require.NoError(t, err, "failed to generate code for %q", thriftFile)
+
+		newHash, err := dirhash(newPackageDir)
+		require.NoError(t, err, "could not hash %q", newPackageDir)
+
+		if newHash != currentHash {
+			t.Fatalf("Generated code for %q is out of date.", thriftFile)
+		}
+	}
+}
+
+func thriftrw(args ...string) error {
```

Comment:
It's just a call to `exec.Command`. Doesn't really need to be part of the library.